File: nodejs/string-compression-443.js
/**
 * @param {character[]} chars
 * @return {number}
 */
'use strict';
var compress = function(chars) {
    let writing = 0;
    let reading = 0;

    for (; reading < chars.length;) {
        let nextReading = reading+1;
        while(nextReading < chars.length && chars[nextReading]===chars[reading]) {
            nextReading++;
        }
        if (nextReading-reading===1) {
            chars[writing] = chars[reading];
            writing++;
        }
        else {
            chars[writing] = chars[reading];
            writing++;
            
            let toWrite = String(nextReading-reading);
            // console.log(toWrite);
            for (let nextWriting=writing;nextWriting-writing < toWrite.length; nextWriting++) {
                chars[nextWriting] = toWrite.charAt(nextWriting-writing);
                // console.log('hello!');
            }
            // chars[writing] = String(nextReading-reading);
            writing = writing + toWrite.length;
        }
        reading = nextReading;
    }
    return writing;
};

// let chars  = ["m","m","m","s","x","x"];
let chars  = ["m","m","m","m","m","m","m","m","m","m","m","m","m","m","m"];
console.log(chars);
let b  = compress(chars);
console.log(b,chars);

File: golang/stringcompression443/string-compression-443.go
package main

import "fmt"

func main() {
	fmt.Println("slaam!")
	a := 21
	b := fmt.Sprintf("%d", a)
	fmt.Println(b)

	chars := [...]byte{'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a'}
	sChars := string(chars[:])
	fmt.Println(sChars)
	res := compress(chars[:])
	ssChars := string(chars[:])
	fmt.Println(ssChars)
	fmt.Println(res)
}
func compress(chars []byte) int {
	reading, writing := 0, 0
	for reading < len(chars) {
		nextReading := reading + 1
		for nextReading < len(chars) && chars[nextReading] == chars[reading] {
			nextReading++
		}
		if nextReading-reading == 1 {
			chars[writing] = chars[reading]
			writing++
		} else {
			chars[writing] = chars[reading]
			writing++
			toWrite := fmt.Sprintf("%d", nextReading-reading)
			for nextWriting := writing; nextWriting-writing < len(toWrite); nextWriting++ {
				chars[nextWriting] = toWrite[nextWriting-writing]
			}
			writing = writing + len(toWrite)
		}
		reading = nextReading
	}
	return writing
}
